### PR TITLE
Validate custom field values on login

### LINKED ISSUE #668: Enforce custom field values against custom data on login form
If you add these custom field definitions to `stormpath.properties` in the case of servlet or `application.properties` in the case of Spring:

```
stormpath.web.login.form.fields.CustomField.type = text
stormpath.web.login.form.fields.CustomField.label = Custom Field Login
stormpath.web.login.form.fields.CustomField.enabled = true
stormpath.web.login.form.fields.CustomField.visible = true
stormpath.web.login.form.fields.CustomField.required = true
stormpath.web.login.form.fields.CustomField.placeholder = Custom Field Login

stormpath.web.register.form.fields.CustomField.type = text
stormpath.web.register.form.fields.CustomField.label = Custom Registration Field
stormpath.web.register.form.fields.CustomField.enabled = true
stormpath.web.register.form.fields.CustomField.visible = true
stormpath.web.register.form.fields.CustomField.required = true
stormpath.web.register.form.fields.CustomField.placeholder = Custom Registration Field
```

You will find that on registration, the custom field value will be stored as custom data.

On login, the custom field will be on the login form, but it's value is *not* enforced against the value in custom data.

### PR DESCRIPTION
**Pull request by mraible**
Fixes #668.

To UAT, I created `examples/servlet/src/main/resources/stormpath.properties` and started Tomcat using `mvn tomcat7:run`. Here's what I put in the properties file:

```
stormpath.web.login.form.fields.favoriteColor.type = text
stormpath.web.login.form.fields.favoriteColor.label = Favorite Color
stormpath.web.login.form.fields.favoriteColor.enabled = true
stormpath.web.login.form.fields.favoriteColor.visible = true
stormpath.web.login.form.fields.favoriteColor.required = true
stormpath.web.login.form.fields.favoriteColor.placeholder = What's your favorite color?

stormpath.web.register.form.fields.favoriteColor.type = text
stormpath.web.register.form.fields.favoriteColor.label = Favorite Color
stormpath.web.register.form.fields.favoriteColor.enabled = true
stormpath.web.register.form.fields.favoriteColor.visible = true
stormpath.web.register.form.fields.favoriteColor.required = true
stormpath.web.register.form.fields.favoriteColor.placeholder = What's your favorite color?
```

I registered with a favorite color of "red". Then I tried logging in with a value of "blue" and confirmed a validation error was displayed. I then entered "red" and verified I was able to login successfully.

screen shot 2016-06-03 at 12 15 48 pm

### CODE REVIEW

File: extensions/servlet/src/main/resources/com/stormpath/sdk/servlet/i18n.properties [hunk 1]
```diff
@@ -44,6 +44,7 @@ stormpath.web.login.form.status.verified = Your Account Has Been Verified. You m
 stormpath.web.login.form.status.created = Your Account Has Been Created. You may now login.
 stormpath.web.login.form.status.forgot = Password Reset Requested. If an account exists for the email provided, you will receive an email shortly.
 stormpath.web.login.form.status.reset = Password Reset Successfully. You can now login with your new password.
+stormpath.web.login.form.customData.validation = The value for "{0}" does not match the value you previously entered. Please try again.
```
Comment (mrioan) [inline extensions/servlet/src/main/resources/com/stormpath/sdk/servlet/i18n.properties]: Just for security reasons I would not provide more information than what we currently have. I would simply replace `Invalid username or password` with `Invalid login information`

Comment (mrioan) [inline extensions/servlet/src/main/resources/com/stormpath/sdk/servlet/i18n.properties]: We need to discuss with the team a way to allow the custom data to be reset. If the user forgets the password he can reset it, but if he forgets the custom data he will not be able to access his account in any way.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/FormController.java [hunk 2]
```diff
@@ -262,4 +264,32 @@ protected void validate(HttpServletRequest request, HttpServletResponse response
...
+     * @param form the form
+     * @return a map of fieldNames and values
+     */
+    protected Map<String, Object> getCustomData(HttpServletRequest request, Form form) {
+        //Custom fields are either declared as form fields which shouldn't not be account fields or through a customField attribute
+        Map<String, Object> result = new LinkedHashMap<String, Object>();
+
+        for (Field field : form.getFields()) {
+            //Field is not part of the default account properties then is a custom field
+            if (DefaultAccount.PROPERTY_DESCRIPTORS.get(field.getName()) == null &&
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/FormController.java]: `DefaultAccount.PROPERTY_DESCRIPTORS` cannot be public. Please see this thread I had with @jarias. Please do revert that change: https://github.com/stormpath/stormpath-sdk-java/pull/678/files. If you have problems we can analyze this together but making the PROPERTY_DESCRIPTORS public should not be considered a valid option here. Thanks @mraible

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/LoginController.java [hunk 3]
```diff
@@ -117,6 +121,22 @@ protected ViewModel onValidSubmit(HttpServletRequest req, HttpServletResponse re
...
+
+        if (account != null) {
+            CustomData customData = account.getCustomData();
+            // validate values in login form match values in custom data
+            Map<String, Object> fields = getCustomData(req, form);
+            for (String fieldName : fields.keySet()) {
+                if (!form.getFieldValue(fieldName).equals(customData.get(fieldName))) {
+                    Field field = form.getField(fieldName);
+                    String fieldLabel = i18n(req, field.getLabel(), field.getLabel());
+                    String msg = i18n(req, "stormpath.web.login.form.customData.validation", new Object[]{fieldLabel});
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/LoginController.java]: Please create a private static final properly for this: `stormpath.web.login.form.customData.validation`

**Comment by mrioan** [comment]
Back to you @mraible

**Comment by jarias** [comment]
According to spec 

> Additional configured login form parameters will be ignored by the framework integration, but should be exposed to the application via the pre and post login handlers.

So this ticket is not actually valid

**Comment by jarias** [comment]
We are keeping the branch just as reference

**Pull request closed without merging**